packages/middleware/src/index.ts

```diff
@@ -73,14 +86,249 @@ export async function build() {
     },
   };
   const middlewareManifestData = JSON.stringify(middlewareManifest, null, 2);
-  const middlewareManifestPath = '.output/server/middleware-manifest.json';
+  const middlewareManifestPath = join(
+    workPath,
+    '.output/server/middleware-manifest.json'
+  );
   await fsp.writeFile(middlewareManifestPath, middlewareManifestData);
 }
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export async function runDevMiddleware(req: IncomingMessage) {
-  // Should run the middleware in the `vm` sandbox and return the result
-  // back to `vercel dev`. If no middleware file exists then this function
-  // should return `finished: false` (very quickly, since this is being
-  // invoked for every HTTP request!).
+const stringifyQuery = (req: IncomingMessage, query: ParsedUrlQuery) => {
+  const initialQueryValues = Object.values((req as any).__NEXT_INIT_QUERY);
+
+  return stringifyQs(query, undefined, undefined, {
+    encodeURIComponent(value: any) {
+      if (initialQueryValues.some(val => val === value)) {
+        return encodeURIComponent(value);
+      }
+      return value;
+    },
+  });
+};
+
+// eslint-disable-next-line
+async function runMiddlewareCatchAll(
+  req: IncomingMessage,
+  res: ServerResponse,
+  requestId: string,
+  name: string,
+  path: string
+) {
+  let result: FetchEventResult | null = null;
+  const parsedUrl = parseUrl(req.url!, true);
+  try {
+    result = await runMiddleware({
+      request: req,
+      response: res,
+      name: name,
+      path,
+      requestId: requestId,
+      parsedUrl,
+      parsed: parseUrl(req.url!, true),
+    });
+  } catch (err) {
+    // if (isError(err) && err.code === 'ENOENT') {
+    //   await this.render404(req, res, parsed)
+    //   return { finished: true }
+    // }
+
+    // const error = isError(err) ? err : new Error(err + '')
+    // console.error(error)
+    // TODO - pretty sure we'll need this.
+    // res.statusCode = 500;
+    // this.sendError(req, res, requestId, 'error in middleware', 500);
+    return { finished: true };
+  }
+
+  if (result === null) {
+    return { finished: true };
+  }
+
+  if (
+    !result.response.headers.has('x-middleware-rewrite') &&
+    !result.response.headers.has('x-middleware-next') &&
+    !result.response.headers.has('Location')
+  ) {
+    result.response.headers.set('x-middleware-refresh', '1');
+  }
+
+  result.response.headers.delete('x-middleware-next');
+
+  // @ts-ignore
+  for (const [key, value] of result.response.headers.entries()) {
+    if (key !== 'content-encoding') {
```

Comment:
How come `content-encoding` is special-cased here?

Reply:
Good question, this was pulled from next.js:
https://github.com/vercel/next.js/blob/canary/packages/next/server/next-server.ts#L1175

Looks like there was an update to the for loop as well that'll need to be pulled in.